dashboard/app/helpers/aichat_openai_helper.rb

```diff
@@ -0,0 +1,48 @@
+module AichatOpenaiHelper
+  def self.get_openai_assistant_response(aichat_model_customizations, stored_messages, new_message, level_id)
+    messages = format_messages(
+      aichat_model_customizations,
+      stored_messages,
+      new_message,
+      level_id
+    )
+
+    request_chat_completion(
+      messages,
+      aichat_model_customizations['temperature'],
+      SharedConstants::AICHAT_MODEL_VERSION
+    )
+  end
+
+  def self.format_messages(aichat_model_customizations, stored_messages, new_message, level_id)
+    level_system_prompt = Level.find_by(id: level_id)&.properties&.dig('aichat_settings', 'levelSystemPrompt') || ""
+    instructions = get_instructions(
+      aichat_model_customizations['systemPrompt'],
+      level_system_prompt,
+      aichat_model_customizations['retrievalContexts']
+    )
+
+    [
+      {role: "system", content: instructions},
+      *stored_messages.map {|message| {role: message['role'], content: message['chatMessageText']}},
+      {role: 'user', content: new_message['chatMessageText']}
+    ]
+  end
+
+  def self.request_chat_completion(messages, temperature, model_version)
+    http_response = OpenaiChatHelper.request_chat_completion(
+      messages,
+      temperature,
+      model_version
+    )
+    JSON.parse(http_response.body)['choices'][0]['message']['content']
+  end
+
+  def self.get_instructions(system_prompt, level_system_prompt, retrieval_contexts)
+    instructions = ""
+    instructions = level_system_prompt + " " unless level_system_prompt.empty?
+    instructions << (system_prompt + " ") unless system_prompt.empty?
+    instructions << retrieval_contexts.join(" ") if retrieval_contexts
+    instructions
+  end
```

Comment:
Yeah, seems like there's probably some opportunity to factor out/share logic between this and the sagemaker implementation (i.e. some common input/output adapters with shared helper functions?) - but agreed with not over-engineering and keeping things separate for now as discussed

Reply:
Yeah totally agree -- this bit is the most egregious (ie, exactly the same as what's in the Sagemaker helper), and it definitely feels like something we should optimize down the line.